File: py/aoc2022/day16.py
"""
Day 16: Proboscidea Volcanium
"""

import heapq
import itertools
import re
from collections import defaultdict
from dataclasses import dataclass

PATTERN = re.compile(
    r"Valve (\w+) has flow rate=(\d+); "
    r"(?:tunnel leads to valve|tunnels lead to valves) (\w+(?:, \w+)*)"
)

SAMPLE_INPUT = [
    "Valve AA has flow rate=0; tunnels lead to valves DD, II, BB",
    "Valve BB has flow rate=13; tunnels lead to valves CC, AA",
    "Valve CC has flow rate=2; tunnels lead to valves DD, BB",
    "Valve DD has flow rate=20; tunnels lead to valves CC, AA, EE",
    "Valve EE has flow rate=3; tunnels lead to valves FF, DD",
    "Valve FF has flow rate=0; tunnels lead to valves EE, GG",
    "Valve GG has flow rate=0; tunnels lead to valves FF, HH",
    "Valve HH has flow rate=22; tunnel leads to valve GG",
    "Valve II has flow rate=0; tunnels lead to valves AA, JJ",
    "Valve JJ has flow rate=21; tunnel leads to valve II",
]


def _parse(lines):
    return {
        (match := re.match(PATTERN, line)).group(1): (
            int(match.group(2)),
            match.group(3).split(", "),
        )
        for line in lines
    }


def _distances(adj):
    keys, distances = set(), defaultdict(lambda: float("inf"))
    for src, dsts in adj:
        keys.add(src)
        distances[src, src] = 0
        for dst, weight in dsts:
            keys.add(dst)
            distances[dst, dst] = 0
            distances[src, dst] = weight
    for mid in keys:
        for src in keys:
            for dst in keys:
                distance = distances[src, mid] + distances[mid, dst]
                if distance < distances[src, dst]:
                    distances[src, dst] = distance
    return distances


@dataclass(order=True, frozen=True)
class _State:
    rooms: tuple[tuple[str, int]]
    valves: frozenset[str]
    flow: int
    total: int
    time: int


def _solve(lines, num_agents, total_time):
    # pylint: disable=too-many-branches,too-many-nested-blocks,too-many-locals
    graph = _parse(lines)
    distances = _distances(
        (src, ((dst, 1) for dst in dsts)) for src, (_, dsts) in graph.items()
    )
    seen, max_seen = set(), 0
    heap = [
        (
            0,
            _State(
                rooms=(("AA", 0),) * num_agents,
                valves=frozenset(src for src, (flow, _) in graph.items() if flow > 0),
                flow=0,
                total=0,
                time=total_time,
            ),
        )
    ]

    while heap:
        estimate, state = heapq.heappop(heap)
        estimate = -estimate
        if state in seen:
            continue
        seen.add(state)
        potential = estimate + sum(
            max(
                (
                    graph[valve][0] * (state.time - delta - 1)
                    for room, age in state.rooms
                    if (delta := distances[room, valve] - age) in range(state.time)
                ),
                default=0,
            )
            for valve in state.valves
        )
        if estimate > max_seen:
            max_seen = estimate
        if potential < max_seen:
            continue

        moves_by_time = defaultdict(lambda: defaultdict(list))
        for valve in state.valves:
            for i, (room, age) in enumerate(state.rooms):
                delta = distances[room, valve] - age
                if delta in range(state.time):
                    moves_by_time[delta][i].append(valve)
        if not moves_by_time:
            continue

        for delta, moves_by_agent in moves_by_time.items():
            for size in range(1, num_agents + 1):
                for combo in itertools.combinations(moves_by_agent.items(), size):
                    indices = [i for i, _ in combo]
                    for valves in itertools.product(*(valves for _, valves in combo)):
                        if len(set(valves)) != size:
                            continue
                        new_rooms = [
                            (room, age + delta + 1) for room, age in state.rooms
                        ]
                        for i, valve in zip(indices, valves):
                            new_rooms[i] = valve, 0
                        rate = sum(graph[valve][0] for valve in valves)
                        new_state = _State(
                            rooms=tuple(sorted(new_rooms)),
                            valves=state.valves - set(valve for valve in valves),
                            flow=state.flow + rate,
                            total=state.total + state.flow * (delta + 1),
                            time=state.time - delta - 1,
                        )
                        heapq.heappush(
                            heap, (-estimate - rate * new_state.time, new_state)
                        )

    return max_seen


def part1(lines):
    """
    >>> part1(SAMPLE_INPUT)
    1651
    """
    return _solve(lines, num_agents=1, total_time=30)


def part2(lines):
    """
    >>> part2(SAMPLE_INPUT)
    1707
    """
    return _solve(lines, num_agents=2, total_time=26)


parts = (part1, part2)

File: kt/src/commonMain/kotlin/com/github/ephemient/aoc2022/Day16.kt
package com.github.ephemient.aoc2022

@Day
class Day16(lines: List<String>) {
    private val graph = lines.associate { line ->
        val (src, rate, dsts) = requireNotNull(PATTERN.matchEntire(line)) { line }.destructured
        src to IndexedValue(rate.toInt(), dsts.split(", "))
    }
    private val distances = run {
        val vertices = buildSet {
            addAll(graph.keys)
            for (value in graph.values) addAll(value.value)
        }
        val vertexIndices = vertices.withIndex().associateBy(keySelector = { it.value }, valueTransform = { it.index })
        val adj = Array(vertices.size) { arrayOfNulls<Int>(vertices.size).apply { set(it, 0) } }
        graph.values.forEachIndexed { i, value -> for (dst in value.value) adj[i][vertexIndices.getValue(dst)] = 1 }
        repeat(vertices.size) { k ->
            repeat(vertices.size) { i ->
                repeat(vertices.size) { j ->
                    val a = adj[i][k]
                    val b = adj[k][j]
                    if (a != null && b != null) adj[i][j] = adj[i][j]?.let { minOf(it, a + b) } ?: (a + b)
                }
            }
        }
        buildMap {
            vertices.forEachIndexed { i, u ->
                this[u] = buildMap {
                    vertices.forEachIndexed { j, v ->
                        adj[i][j]?.let { this[v] = it }
                    }
                }
            }
        }
    }

    @Day.Part
    fun part1(): Int = solve(1, 30)

    @Day.Part
    fun part2(): Int = solve(2, 26)

    @Suppress("CyclomaticComplexMethod", "LongMethod", "LoopWithTooManyJumpStatements", "NestedBlockDepth")
    private fun solve(agents: Int, totalTime: Int): Int {
        var max = 0
        val seen = mutableSetOf<State>()
        val queue = PriorityQueue<IndexedValue<State>>(compareBy(reverseOrder()) { it.index })
        queue.add(
            IndexedValue(
                0,
                State(
                    rooms = List(agents) { IndexedValue(0, "AA") },
                    valves = graph.entries.mapNotNullTo(mutableSetOf()) { (key, value) ->
                        if (value.index > 0) key else null
                    },
                    flow = 0,
                    total = 0,
                    timeRemaining = totalTime,
                )
            )
        )
        while (!queue.isEmpty()) {
            val (estimate, state) = queue.remove()
            if (!seen.add(state)) continue
            val potential = estimate + state.valves.sumOf { valve ->
                state.rooms.maxOfOrNull { (age, room) ->
                    val d = (distances[room]?.get(valve) ?: return@maxOfOrNull 0) - age
                    if (d in 0 until state.timeRemaining) {
                        graph.getValue(valve).index * (state.timeRemaining - d - 1)
                    } else {
                        0
                    }
                } ?: 0
            }
            if (estimate > max) max = estimate
            if (potential < max) continue
            val movesByTime: Map<Int, List<List<String>>> = buildMap<_, MutableList<MutableList<String>>> {
                for (valve in state.valves) {
                    state.rooms.forEachIndexed { i, room ->
                        val d = (distances[room.value]?.get(valve) ?: return@forEachIndexed) - room.index
                        if (d in 0 until state.timeRemaining) {
                            getOrPut(d) { MutableList(agents) { mutableListOf() } }[i].add(valve)
                        }
                    }
                }
            }
            if (movesByTime.values.all { movesByAgent -> movesByAgent.all { it.isEmpty() } }) continue
            for ((d, movesByAgent) in movesByTime) {
                val indices = IntArray(agents) { -1 }
                while (
                    indices.withIndex().any { (i, j) ->
                        (if (j < movesByAgent[i].lastIndex) j + 1 else -1).also { indices[i] = it } >= 0
                    }
                ) {
                    val allUnique: Boolean
                    val valves = buildSet {
                        allUnique = indices.withIndex().all { (i, j) -> j < 0 || add(movesByAgent[i][j]) }
                    }
                    if (allUnique) {
                        val rooms = indices.mapIndexed { i, j ->
                            if (j >= 0) {
                                IndexedValue(0, movesByAgent[i][j])
                            } else {
                                val (age, room) = state.rooms[i]
                                IndexedValue(age + d + 1, room)
                            }
                        }.sortedWith(compareBy({ it.value }, { it.index }))
                        val rate = indices.withIndex().sumOf { (i, j) ->
                            if (j >= 0) graph.getValue(movesByAgent[i][j]).index else 0
                        }
                        val newState = State(
                            rooms = rooms,
                            valves = state.valves - valves,
                            flow = state.flow + rate,
                            total = state.total + state.flow * (d + 1),
                            timeRemaining = state.timeRemaining - d - 1,
                        )
                        queue.add(IndexedValue(estimate + rate * newState.timeRemaining, newState))
                    }
                }
            }
        }
        return max
    }

    private data class State(
        val rooms: List<IndexedValue<String>>,
        val valves: Set<String>,
        val flow: Int,
        val total: Int,
        val timeRemaining: Int,
    )
}

private val PATTERN =
    """Valve (\w+) has flow rate=(\d+); (?:tunnel leads to valve|tunnels lead to valves) (\w+(?:, \w+)*)""".toRegex()
